### DISCUSSION TITLE
Copy the command output between the prompts

### DISCUSSION THREAD

**Discussion by ymotongpoo** [wezterm/wezterm#2343] [Q&A] [answered]
Hi all, I've recently switched to WezTerm and I'm totally enjoying using it. Thank you for the awesome product and great community around it.

Now I'm trying enrich my WezTerm experience with the custom feature to copy the text between the last prompt to the current prompt. This is pretty useful when I write an operation logs, such as bug report and runbook documentations.

Though I have searched the way to fetch buffered text data from pane, I only could find [`Pane.get_line_as_text()`](https://wezfurlong.org/wezterm/config/lua/pane/get_lines_as_text.html) and [`Pane.get_logical_lines_as_text()`](https://wezfurlong.org/wezterm/config/lua/pane/get_logical_lines_as_text.html), so I supposed that is the only functions to get texts from pane. These should work when the command output is short enough and inside the viewport, like the way shared [here](https://github.com/wez/wezterm/discussions/628#discussion-3304970), but in the case the command output is long enough and multiple viewports long, it doesn't work.

I would like to get any ideas on how to get text between the prompts. The best case is to fetch the text between `OSC "133;C"` and `OSC "133;D"` ([reference](https://gitlab.freedesktop.org/Per_Bothner/specifications/blob/master/proposals/semantic-prompts.md))

I was wondering if it's possible for Lua API to:

1. Activate copy mode (`wezterm.action.ActivateCopyMode`)
2. Mark the start of the selection to the previous prompt (combination of `wezterm.action.ScrollToPrompt(-1)` and something that moves the cursor to the top of the viewport and mark selection start)
3. Mark the end of the selection to the current prompt (combination of `wezterm.action.ScrollToPrompt(1)` and something that moves the cursor to the top of the viewport and mark selection end)
4. Copy the selection to clipboard (maybe `wezterm.action.CopyTo 'Clipboard'`)

but I couldn't find the way to operate cursor in CopyMode.

**Comment by wez** [comment, marked as answer]
Hmm, I haven't tried to automate this or do it completely without the mouse so far. I use https://wezfurlong.org/wezterm/config/lua/keyassignment/SelectTextAtMouseCursor.html to grab the semantic zone when I triple click:

```lua
local wezterm = require 'wezterm'

return {
  mouse_bindings = {
    {
      event = { Down = { streak = 3, button = 'Left' } },
      action = wezterm.action.SelectTextAtMouseCursor 'SemanticZone',
      mods = 'NONE',
    },
  },
}
```

I think it would be a reasonable request to expand copy mode to support something like `act.CopyMode { SetSelectionMode = 'SemanticZone' }` as well as `act.CopyMode 'MoveForwardSemanticZone` and `act.CopyMode 'MoveBackwardSemanticZone'` and/or `Move(Backward|Forward)Prompt` or similar to specify what kind of semantic zone you want to move to.

**Reply by ymotongpoo** [reply]
Thank you @wez for the quite spot on answer as the work around for my current demand. Also I appreciate your confirming that the operations I was trying to achieve is not possible with the current Lua API.

Definitely I will post a request regarding semantic zones. Again, thank you so much!

**Reply by rkabrick** [reply]
Does this answer remain the best answer? Setting this in my mouse bindings simply copies the entirety of text in the terminal, not just the most recent block. If that is what is meant by semantic zone that may just be my own misunderstanding

**Reply by cassepipe** [reply]
@rkabrick 
It is working for me on wezterm 20240203-110809-5046fc22 on linux so there might be something wrong on your end

**Reply by sadamczyk** [reply]
@rkabrick That sounds like Wezterm isn't recognizing the semantic zones correctly.

You need to setup shell integration first, otherwise Wezterm won't know the boundaries of your prompt, input/outputs.
See https://wezterm.org/shell-integration.html

**Reply by saumyajyoti** [reply]
How to copy prompt input text via action? The above solution copes the output text zone.

**Comment by MartenBE** [comment]
Stumled on this topic, I wrote this to save the last command and output to a file:

```lua
wezterm.on("save-output", function(window, pane)
	local zones = pane:get_semantic_zones()
	local i = #zones
	local last_index = nil
	while i > 0 and not last_index do
		if zones[i].semantic_type == "Output" then
			last_index = i
		end
		i = i - 1
	end

	if not last_index then
		local err_msg = "No output found in the current pane."
		wezterm.log_error(err_msg)
		window:toast_notification("Error", err_msg)
		return
	end

	local latest_output_zone = zones[last_index]
	local latest_output_text = pane:get_text_from_semantic_zone(latest_output_zone)
	local latest_input_zone = zones[last_index - 1]
	local latest_input_text
	if latest_input_zone then
		latest_input_text = pane:get_text_from_semantic_zone(latest_input_zone)
	else
		latest_input_text = "[No input found]"
	end

	local text = latest_input_text .. "\n\n" .. latest_output_text

	local cwd_uri = pane:get_current_working_dir()
	if not cwd_uri or cwd_uri.scheme ~= "file" then
		local err_msg = "Failed to get current working directory. URI: " .. tostring(cwd_uri)
		wezterm.log_error(err_msg)
		window:toast_notification("Error", err_msg)
		return
	end

	local output_file_path = cwd_uri.file_path .. "/wezterm-output-" .. os.date("%Y%m%d-%H%M%S") .. ".log"
	local output_file, err = io.open(output_file_path, "w")
	if not output_file then
		local err_msg = "Failed to open file " .. output_file_path .. " for writing: " .. err
		wezterm.log_error(err_msg)
		window:toast_notification("Error", err_msg)
		return
	end

	output_file:write(text)
	output_file:flush()
	output_file:close()

	local info_msg = "Wrote output to file: " .. output_file_path
	wezterm.log_info(info_msg)
	window:toast_notification("Info", info_msg, tostring(cwd_uri), 5000)
end)
```

```lua
config.leader = { key = "a", mods = "CTRL" }
config.keys = {
	...
	{ key = "o", mods = "LEADER", action = wezterm.action({ EmitEvent = "save-output" }) },
}
```

Doesn't work well in an SSH session (copies everything) though

**Reply by ryanthomas-org** [reply]
Thanks so much for putting this together. I used your example to create the following which is similar, as I am brought to this page when searching for what I want. I have become accustomed to having a shortcut key that simply copies the output of the last command run to the clipboard. I am not interested in getting a copy of the command I ran sent back to me so I was able to trim this greatly. 

```
wezterm.on("save-output", function(window, pane)

    --https://github.com/wezterm/wezterm/discussions/2343
    local zones = pane:get_semantic_zones()
    local i = #zones
    local last_index = nil
    while i > 0 and not last_index do
        if zones[i].semantic_type == "Output" then
            last_index = i
        end
        i = i - 1
    end

    if not last_index then return nil end 
    local latest_output_zone = zones[last_index]
    local text = pane:get_text_from_semantic_zone(latest_output_zone)
    if text == nil then return nil end
    io.popen('xclip -selection clipboard','w'):write(text):close()

end)

config.keys = {

    { key = "o", mods = "ALT", action = wezterm.action({ EmitEvent = "save-output" }) },
}
```
This is enough that I can now begin experimenting with wezterm as my daily driver. :) 

EDIT 15 NOV 2025: The above was failing intermittently; it was tough to replicate on demand and therefore hard to troubleshoot, but the following version is working more consistently now:

```
local wezterm = require("wezterm")

wezterm.on("save-output", function(window, pane)

    local text
    local zones = pane:get_semantic_zones()

    for i = 1, #zones do 
        local z = zones[i]
        local text_match = pane:get_text_from_semantic_zone(z)
        if z.semantic_type == "Output" and string.find(text_match, ".") then
            text = text_match
        end
    end

    if text == nil then return nil end
    io.popen('xclip -selection clipboard','w'):write(text):close()

end)

config.keys = {

    { key = "o", mods = "ALT", action = wezterm.action({ EmitEvent = "save-output" }) },
}
```
